src/v/kafka/server/group_manager.cc

```diff
@@ -564,6 +578,42 @@ group_manager::leave_group(leave_group_request&& r) {
     }
 }
 
+ss::future<txn_offset_commit_response>
+group_manager::txn_offset_commit(txn_offset_commit_request&& r) {
+    if (!_catchup_lock.take_reader_lock()) {
+        vlog(klog.warn, "can't process a tx: coordinator_load_in_progress");
+        co_return txn_offset_commit_response(
+          r, error_code::coordinator_load_in_progress);
+    }
+
+    try {
+        auto error = validate_group_status(
+          r.ntp, r.data.group_id, offset_commit_api::key);
+        if (error != error_code::none) {
+            _catchup_lock.release_reader_lock();
+            co_return txn_offset_commit_response(r, error);
+        }
+
+        auto group = get_group(r.data.group_id);
+        if (!group) {
+            // <kafka>the group is not relying on Kafka for group management, so
+            // allow the commit</kafka>
+            auto p = _partitions.find(r.ntp)->second->partition;
+            group = ss::make_lw_shared<kafka::group>(
+              r.data.group_id, group_state::empty, _conf, p);
+            _groups.emplace(r.data.group_id, group);
+            _groups.rehash(0);
+        }
+
+        auto reply = co_await group->handle_txn_offset_commit(std::move(r));
+        _catchup_lock.release_reader_lock();
```

Comment:
can you actually create a struct, that takes a pointer and in the destructor it always calls `_catchup_lock.release_reader_lock()` I think seastar might have a `lock_guard` see the batch_cache `lock_guard` for example. 

then we would not need the try-catch at all.

Reply:
when migrated to `seastar::lock` you can use `ss::get_units` they implement the RAII unlock

Reply:
@mmaslankaprv [rwlock](http://docs.seastar.io/master/classseastar_1_1basic__rwlock.html) doesn't have get_units can you expand your thought please?

Reply:
I think I found, there is [hold_read_lock](http://docs.seastar.io/master/rwlock_8hh_source.html) do you know how RAII works with co_await / co_return I think I should ensure somehow that `u` is passed to the continuation to prevent premature destruction, shouldn't I?